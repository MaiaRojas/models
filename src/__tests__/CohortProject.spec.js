const mongoose = require('mongoose');
const {
  CohortProject,
  Cohort,
  Project,
  Campus,
} = require('../../')(mongoose);

const campusJSON = {
  slug: 'lim',
  name: 'Lima',
  locale: 'es-PE',
  timezone: 'America/Lima',
  active: true,
};

const projectJSON = {
  slug: 'cipher',
  repo: 'Laboratoria/curricula-js',
  path: 'projects/01-cipher',
  version: '2.0.0',
  parserVersion: '1.1.1',
  createdAt: new Date(),
  prefix: 1,
  title: 'Cifrado CÃ©sar',
  rubric: '2',
  locale: 'es-ES',
  track: 'js',
  skills: {},
};

describe('CohortProject', () => {
  beforeAll(async () => {
    await mongoose.connect(global.__MONGO_URI__, { useNewUrlParser: true });
  });

  afterAll(async () => {
    await mongoose.disconnect();
  });

  beforeEach(async () => {
    await mongoose.connection.db.dropDatabase();
    await CohortProject.createIndexes();
    await Cohort.createIndexes();
    await Project.createIndexes();
  });

  it('should fail when missing props', () => {
    const cohortProject = new CohortProject();
    return cohortProject.save()
      .catch(err => expect(err.errors).toMatchSnapshot());
  });

  it('should fail when cohort or project are not ObjecIds', () => {
    const doc = new CohortProject({
      cohort: 'cohort._id',
      project: 'project._id',
    });
    return doc.save()
      .catch(err => expect(err.errors).toMatchSnapshot());
  });

  it('should fail when cohort does not exist', async () => {
    const project = new Project(projectJSON);
    const doc = new CohortProject({
      project: project._id,
      cohort: (new Cohort())._id,
    });

    await project.save();

    return doc.save()
      .catch((err) => {
        expect(err.message).toBe('Cohort does not exist');
      });
  });

  it('should fail when project does not exist', async () => {
    const campus = new Campus(campusJSON);
    const cohort = new Cohort({
      campus: campus._id,
      program: 'bc',
      track: 'core',
      generation: 1,
      start: new Date(),
      end: new Date(),
    });
    const doc = new CohortProject({
      cohort: cohort._id,
      project: (new Project())._id,
    });

    await campus.save();
    await cohort.save();

    return doc.save()
      .catch((err) => {
        expect(err.message).toBe('Project does not exist');
      });
  });

  it('should save successfully with appropriate fields', () => {
    const cohort = new Cohort({});
    const project = new Project({});
    const cohortProject = new CohortProject({
      cohort: cohort._id,
      project: project._id,
    });
    return cohortProject.save()
      .catch(err => expect(err.errors).toMatchSnapshot());
  });

  it('should save successfully with appropriate fields', async () => {
    const campus = new Campus(campusJSON);
    const cohort = new Cohort({
      campus: campus._id,
      program: 'bc',
      track: 'core',
      generation: 1,
      start: new Date(),
      end: new Date(),
    });

    const project = new Project(projectJSON);
    const cohortProject = new CohortProject({
      cohort: cohort._id,
      project: project._id,
    });

    await campus.save();
    await cohort.save();
    await project.save();

    return cohortProject.save()
      .then((result) => {
        expect(`${result.project}`).toBe(`${project._id}`);
        expect(`${result.cohort}`).toBe(`${cohort._id}`);
      });
  });
});
