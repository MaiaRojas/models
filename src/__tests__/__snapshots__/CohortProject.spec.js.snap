// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CohortProject should fail when cohort or project are not ObjecIds 1`] = `
Object {
  "cohort": [CastError: Cast to ObjectID failed for value "cohort._id" at path "cohort"],
  "project": [CastError: Cast to ObjectID failed for value "project._id" at path "project"],
}
`;

exports[`CohortProject should fail when missing props 1`] = `
Object {
  "cohort": [ValidatorError: Path \`cohort\` is required.],
  "project": [ValidatorError: Path \`project\` is required.],
}
`;

exports[`CohortProject should save successfully with appropriate fields 1`] = `undefined`;
